### PR TITLE
Add optional ctl_connection_settings, force failure when can't get Table Metadata

### PR DESCRIPTION
**Pull request by kdparker**
Failure on replication handler when it can't get table metadata looks like this:
https://pb.yelpcorp.com/188948

This branch was tested using branch keegan_DATAPIPE-844-replhandler in the replication handler (where running it with this fix caused DATAPIPE-786 to be fixed without using the bandaid refresh runner)

### CODE REVIEW

File: pymysqlreplication/binlogstream.py [hunk 1]
```diff
@@ -56,14 +57,16 @@ def __init__(self, connection_settings, server_id, resume_stream=False,
...
         self.__connection_settings = connection_settings
-	if not connection_settings.get("charset"):
+        if not connection_settings.get("charset"):
             self.__connection_settings["charset"] = "utf8"
 
         self.__connected_stream = False
         self.__connected_ctl = False
         self.__resume_stream = resume_stream
         self.__blocking = blocking
+        self._ctl_connection_settings = ctl_connection_settings
```
Comment (justincinmd) [inline pymysqlreplication/binlogstream.py:166]: Do you need to set a charset for this too?

File: pymysqlreplication/row_event.py [hunk 2]
```diff
@@ -55,7 +55,7 @@ def __init__(self, from_packet, event_size, table_map, ctl_connection, **kwargs)
         self.columns = self.table_map[self.table_id].columns
 
         if len(self.columns) == 0:  # could not read the table metadata, probably already dropped
-            self.complete = False
+            raise TableMetadataUnavailableError("Table Metadata could not be read for {}".format(self.table))
```
Comment (justincinmd) [inline pymysqlreplication/row_event.py]: Maybe this should be optional?  I absolutely want it, but the maintainer didn't - I think it's much more correct to raise an exception, but if we want to contribute back upstream, we probably shouldn't force the issue.

File: pymysqlreplication/binlogstream.py [hunk 3]
```diff
@@ -35,14 +35,16 @@ class BinLogStreamReader(object):
...
 
-    def __init__(self, connection_settings, server_id, resume_stream=False,
+    def __init__(self, connection_settings, server_id, ctl_connection_settings=None, resume_stream=False,
                  blocking=False, only_events=None, log_file=None, log_pos=None,
                  filter_non_implemented_events=True,
                  ignored_events=None, auto_position=None,
                  only_tables=None, only_schemas=None,
-                 freeze_schema=False, skip_to_timestamp=None):
+                 freeze_schema=False, skip_to_timestamp=None,
+                 fail_on_table_metadata_unavailable=False):
```
Comment (premsantosh) [inline pymysqlreplication/binlogstream.py:137]: Is this a temporary flag? And why is it defaulted to false? When do we not want to get alerted?

File: pymysqlreplication/exceptions.py [hunk 4]
```diff
@@ -0,0 +1,2 @@
+class TableMetadataUnavailableError(Exception):
+    pass
```
Comment (premsantosh) [inline pymysqlreplication/exceptions.py]: Would be nice to have a little context with this. Maybe something like:

```
class TableMetadataUnavailableError(Exception):
  def __init__(self, table_name, cluster):
    Exception.__init__(self, "Unable to find metadata for table {t} in cluster {c}".format(..))
```

**Review by premsantosh — CHANGES_REQUESTED** [review]

File: pymysqlreplication/binlogstream.py (hunk 3 above)
Comment (kdparker) [reply pymysqlreplication/binlogstream.py:137]: I think the intention was we eventually wanted to merge this in to the overall master branch of the library, and to do so we want to keep existing functionality exactly the same (the corresponding branch in replhandler sets this to True).

Do we want to make this default to True?

**Comment by kdparker** [comment]
Something else we may want to consider is migrating this to another user/merge it in to the noplay master (can anyone here actually merge these changes?)

File: pymysqlreplication/binlogstream.py (hunk 3 above)
Comment (premsantosh) [reply pymysqlreplication/binlogstream.py:137]: Aah I see, that makes sense. If we are setting it to true from our side then it should be fine.

**Comment by kdparker** [comment]
Not using this branch anymore, closing

**Pull request closed without merging**